File: 04/1.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func readLines(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func main() {
	file_path := "1.in"
	lines, err := readLines(file_path)
	if err != nil {
		log.Fatal(err)
	}

	m := len(lines)
	n := len(lines[0])

	ans := 0

	target := "XMAS"
	for x := range m {
		for y := range n {
			if lines[x][y] != 'X' {
				continue
			}
			for i := -1; i <= 1; i += 1 {
				for j := -1; j <= 1; j += 1 {
					if i == 0 && j == 0 {
						continue
					}
					f := true
					dx := x
					dy := y
					for k := range 3 {
						dx = dx + i
						dy = dy + j
						if dx < m && dx >= 0 && dy < n && dy >= 0 {
							if lines[dx][dy] != target[k+1] {
								f = false
								break
							}
							fmt.Printf("cur %c target %c\t", lines[dx][dy], target[k+1])

						} else {
							f = false
							break
						}
					}
					if f {
						ans += 1
					}
					fmt.Println()
				}
			}
		}
	}

	fmt.Println(ans)
}

File: 08/1.jl
using Printf

f = open("1.in", "r")
lines = readlines(f)
close(f)

m = length(lines)
n = length(lines[1])

mp = Dict{Char,Vector{Tuple{Int,Int}}}()
for i in 1:m
    for j in 1:n
        c = lines[i][j]
    if c == '.'
        continue
    end
        if !haskey(mp, c)
            mp[c] = Vector{Tuple{Int,Int}}()
        end
        push!(mp[c], (i, j))
    end
end

grid = [collect(line) for line in lines]

function check!(x, y,grid)
    if x < 1 || x > m || y < 1 || y > n
        return 0
    end
    if grid[x][y] != '%'
        grid[x][y] = '%'
        return 1
    end
    return 0
end

ans = 0
for (key, value) in mp
    if length(value) == 1
      x , y = value[1]
      global ans += 1
    end
    for i in 1:length(value)
        for j in i+1:length(value)
            x1, y1 = value[i]
            x2, y2 = value[j]
            dx = x2 - x1
            dy = y2 - y1

            global ans += check!(x1 - dx, y1 - dy,grid)
            global ans += check!(x2 + dx, y2 + dy,grid)
        end
    end
end

for line in grid
  for c in line
    print(c)
  end
  println()
end
println(ans)
